soc/arm/nuvoton_npcx/common/scfg.c

```diff
@@ -6,6 +6,7 @@
 
 #include <device.h>
```

Comment:
All the code in this file should likely be in a pinctrl driver, what is the point of having this outside of the driver? I understand you had before a custom implementation, but what is the reason to keep this in `soc/` now? The only driver managing pinctrl related registers should be the pinctrl driver itself. Unfortunately, `soc/` tends to be a blackhole with too much custom "internal APIs"  for no reason.

To illustrate the problem:

We have that `scfg.c` (in `soc/`) includes `drivers/pinctrl.h` (as if it was a pinctrl client). APIs in this file get exposed via the internal header `soc_pins.h`, which is actually used by the `pinctrl_npcx.c` (the pinctrl driver) via `soc.h`, closing the "circle". And then we have a pinctrl driver that looks like this:

```c
/*
 * Copyright (c) 2022 Nuvoton Technology Corporation.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#include <assert.h>
#include <drivers/pinctrl.h>
#include <soc.h>

int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
			   uintptr_t reg)
{
	ARG_UNUSED(reg);

	npcx_pinctrl_mux_configure(pins, pin_cnt, 1);

	return 0;
}
```

So, I don't think this “architecture” or way of doing things is acceptable when we have a proper API. I could understand that sort of custom stuff with APIs that are not mature, like it could be clock control, but definitely not here. The purpose of pinctrl driver is to:

1. Configure pinmux settings, e.g., route a peripheral signal to a particular port
2. Allow to configure pin properties for correct operation, e.g., use open-drain mode when using I2C.

GPIO API needs to be used only when pins are controlled manually by the application, that is, no peripheral is driving them. It's true that in some cases we have an overlapping between GPIO and pinctrl drivers, we may improve this in the future, but the split is clear now.

Reply:
Note: the design principles of pinctrl are detailed here, for reference: https://docs.zephyrproject.org/latest/guides/pinctrl/index.html

Reply:
Thanks for the detailed explanation. Understand the disadvantages of exposing soc custom implementations via drivers/pinctrl.h. I will remove soc included files in `pinctrl_soc.h` and move functions related to pinctrl mechanism from `scfg.c` to `pinctrl_npcx.c`.

Reply:
Done in the new PR.